File: benchmarks/db_join/db_join.py
from UTIL import shared


def cross_join(
    A: shared[list[int]],
    Len_A: int,
    Att_A: int,
    B: shared[list[int]],
    Len_B: int,
    Att_B: int,
) -> shared[list[int]]:
    ret: list[int] = []

    Att = Att_A + Att_B - 1

    for _ in range(0, Len_A * Len_B * Att + 1):
        ret = ret + [0]

    ret_idx = 0

    for i in range(0, Len_A):
        for j in range(0, Len_B):
            if A[i * Att_A] == B[j * Att_B]:
                ret[ret_idx * Att] = A[i * Att_A]
                ret[ret_idx * Att + 1] = A[i * Att_A + 1]
                ret[ret_idx * Att + 2] = B[j * Att_B + 1]
                ret_idx = ret_idx + 1

    return ret


A = [1, 2, 3, 4]
B = [1, 5, 2, 6]
Len_A = 2
Len_B = 2
Att_A = 2
Att_B = 2
print(cross_join(A, Len_A, Att_A, B, Len_B, Att_B))

File: benchmarks/db_join/64/db_join.c
/*
 * DB Cross Join Benchmark
 *
 * Performs a cross join on two relations (A and B) on the first attribute.
 * Each tuple has 2 attributes; output rows have 3 attributes.
 *
 * Party 0 (Server): Owns A
 * Party 1 (Client): Owns B
 *
 * Parameters:
 *   LEN_A = 64
 *   LEN_B = 64
 */

#define LEN_A 64
#define LEN_B 64
#define ATT_A 2
#define ATT_B 2
#define ATT_OUT (ATT_A + ATT_B - 1)
#define OUTPUT_SIZE (LEN_A * LEN_B * ATT_OUT)

typedef struct {
    int result[OUTPUT_SIZE];
} Output;

Output main(
    __attribute__((private(0))) int A[LEN_A * ATT_A],
    __attribute__((private(1))) int B[LEN_B * ATT_B]
)
{
    Output output;
    int ret_idx = 0;

    for (int i = 0; i < OUTPUT_SIZE; i++) {
        output.result[i] = 0;
    }

    for (int i = 0; i < LEN_A; i++) {
        for (int j = 0; j < LEN_B; j++) {
            if (A[i * ATT_A] == B[j * ATT_B]) {
                output.result[ret_idx * ATT_OUT] = A[i * ATT_A];
                output.result[ret_idx * ATT_OUT + 1] = A[i * ATT_A + 1];
                output.result[ret_idx * ATT_OUT + 2] = B[j * ATT_B + 1];
                ret_idx = ret_idx + 1;
            }
        }
    }

    return output;
}
